### PR TITLE
[Heartbeat] Unpack beats to enable capabilities inside container

### PR DESCRIPTION
**Pull request by emilioalvap**
## What does this PR do?

This PR enables unpacking of beats inside the container at build time, so that required `cap_net_raw, cap_setuid` capabilities can be assigned to the binary.

## Why is it important?

Without the required capabilities, heartbeat cannot execute ICMP pings or setuid calls. As it is now, agent is unpacking beats at runtime, most likely with a user that doesn't have permission to assign capabilities.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

- Build elastic-agent containers, from `x-pack/elastic-agent` run:
`env PLATFORMS="+all linux/amd64" mage dev:package`
- Run one of the built containers and provide some heartbeat configuration:
```
docker run --name agent -it -u root --env FLEET_ENROLL=1 --env \
FLEET_URL=<url> --env \ 
FLEET_ENROLLMENT_TOKEN=<token> \ 
docker.elastic.co/beats/elastic-agent:8.1.0-SNAPSHOT
```

## Related issues
- Relates #27651

### CODE REVIEW

File: dev-tools/packaging/packages.yml [hunk 1]
```diff
@@ -482,6 +482,8 @@ shared:
       user: '{{ .BeatName }}'
       linux_capabilities: ''
       image_name: ''
+      unpack_beats: true
```
Comment (andrewvc) [inline dev-tools/packaging/packages.yml:485]: If we're saying this is the default behavior, do we need this option, or should we just always do this?

Comment (blakerouse) [reply dev-tools/packaging/packages.yml:485]: I think we should make this the default and remove this option.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl [hunk 2]
```diff
@@ -181,6 +181,17 @@ RUN mkdir /app
 RUN chown {{ .user }} /app
 {{- end }}
 {{- end }}
+
+{{- if .unpack_beats }}
+RUN mkdir -p {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    for beatPath in {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/downloads/*.tar.gz; do \
+    tar xf $beatPath -C {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    rm $beatPath; \
```
Comment (blakerouse) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:189]: We should not remove this $beatPath. It's not going to save any space in the image anyway because it is included in a previous layer.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl [hunk 3]
```diff
@@ -181,6 +181,17 @@ RUN mkdir /app
...
+    done && \
+    chown -R {{ .user }} {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    setcap cap_net_raw,cap_setuid+p {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }}/heartbeat-*/heartbeat
```
Comment (blakerouse) [inline dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:192]: What is the cap_setuid+p for? Why is that needed?

Comment (emilioalvap) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:192]: Heartbeat requires it for switching user id at runtime

Comment (andrewvc) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:192]: To add some detail here @blakerouse heartbeat runs node as a subprocess, and node hates to run as root. It's also just a best practice to setuid out of root if you can regardless.

Comment (blakerouse) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:192]: Okay. It was just more of a question so I can understand. Thanks for the explanation.

File: dev-tools/packaging/packages.yml (hunk 1 above)
Comment (emilioalvap) [reply dev-tools/packaging/packages.yml]: Ok, removed the option and it's now unpacking by default

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl (hunk 2 above)
Comment (emilioalvap) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl]: True that, removed

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl [hunk 4]
```diff
@@ -181,6 +181,17 @@ RUN mkdir /app
...
 {{- end }}
 {{- end }}
+
+{{- if .unpack_beats }}
+RUN mkdir -p {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    for beatPath in {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/downloads/*.tar.gz; do \
+    tar xf $beatPath -C {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    rm $beatPath; \
+    done && \
+    chown -R {{ .user }} {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
```
[the remark this answers is not in the document]
Comment (emilioalvap) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl]: Group was root, changed that to be `elastic-agent` too.

**Comment by blakerouse** [comment]
@emilioalvap What is the status of this PR? Can we get it up for full review?

We believe this will also fix https://github.com/elastic/observability-dev/issues/1935.

File: dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl [hunk 5]
```diff
@@ -181,6 +181,16 @@ RUN mkdir /app
...
 {{- end }}
 {{- end }}
+
+RUN mkdir -p {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    for beatPath in {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/downloads/*.tar.gz; do \
+    tar xf $beatPath -C {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }}; \
+    done && \
+    chown -R {{ .user }}:{{ .user }} {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }} && \
+    chown -R root:root {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }}/*/*.yml && \
+    chmod 0644 {{ $beatHome }}/data/{{.BeatName}}-{{ commit_short }}/{{ .beats_install_path }}/*/*.yml && \
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:191]: @emilioalvap I think the first workaround is okay, as Elastic Agent never writes this file.

As for the post-build check I see you do a `chmod 0644` for the `*.yml`, can that be changed to `chmod 0600` or would that break it?

[the remark this answers is not in the document]
Comment (emilioalvap) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:191]: @blakerouse Some of the beats' *.yml files come with `0600` and that is actually what breaks the test. `0644` seems to be the permission mask applied to *.yml files outside zipped beats. So now that we are unpacking beats, the test that used to check elastic.-agent *.yml files is also checking other beats config files.

[the remark this answers is not in the document]
Comment (blakerouse) [reply dev-tools/packaging/templates/docker/Dockerfile.elastic-agent.tmpl:191]: @emilioalvap Still confused? So are you saying that its working with the change and the test is passing or we need to change either one?
